Add support for XLM-R XL and XXL models by modeling_xlm_roberta_xl.py

Soonhwan-Kwon opened the pull request:
This PR adds support for the newly released XL and XXL models for XLM-R, and . These models are described in the "Larger-Scale Transformers for Multilingual Masked Language Modeling" paper.

And thank you for @patrickvonplaten and @stefan-it, as the review I got from the #13210, 
I added modeling_xlm_roberta_xl.py and convert_xlm_roberta_xl_original_pytorch_checkpoint_to_pytorch.py for conversion script.

I compared fairseq and transformers side by side,
and managed output same.

torch.Size([1, 11, 250880]) torch.Size([1, 11, 250880])
max_absolute_diff = 0.000186920166015625
Do both models output the same tensors? 🔥
Saving model to converted_xlmr_xl
Configuration saved in converted_xlmr_xl/config.json
Model weights saved in converted_xlmr_xl/pytorch_model.bin

Since fairseq roberta to transformers conversion was made a long time ago,
transformers architecture differs far from fairseq which originally started from,
and it makes quite confusion to write right code.
I synced transformers code to allow fairseq model structure.

- [ ] add test for XLM-R XL and XXL
- [ ] upload model for XLM-R XL and XXL to official repo

Soonhwan-Kwon committed:
add xlm roberta xl

Soonhwan-Kwon committed:
add convert xlm xl fairseq checkpoint to pytorch

Soonhwan-Kwon committed:
fix init and documents for xlm-roberta-xl

Soonhwan-Kwon committed:
fix indention

patrickvonplaten commented:
Thanks for the PR @Soonhwan-Kwon!

Could you also add a test file and some integration tests? :-)

Soonhwan-Kwon committed:
add test for XLM-R xl,xxl

patrickvonplaten commented:
Hey @Soonhwan-Kwon, 

Thanks a lot for working and this and sorry to reply so late!
Would it be ok to upload the checkpoints for now under your name on the hub and to make the tests pass and then in a last step, will move the checkpoints to the official organization?

Let me know if you need some help fixing the last steps :-)

Soonhwan-Kwon committed:
fix model hub name

Soonhwan-Kwon commented:
> Hey @Soonhwan-Kwon,
> 
> Thanks a lot for working and this and sorry to reply so late! Would it be ok to upload the checkpoints for now under your name on the hub and to make the tests pass and then in a last step, will move the checkpoints to the official organization?
> 
> Let me know if you need some help fixing the last steps :-)

Thank you for the reply, I'm middle of uploading models, but it takes time for xxlarge(over 24GB) model.

Soonhwan-Kwon commented:
@patrickvonplaten I have uploaded all models, but I have no idea because I'm kind of newbie here. How can I fix last steps? Thank you in advance.

patrickvonplaten commented:
@Soonhwan-Kwon, could you maybe also add a configuration file (just copy the xlm-roberta one) and also add a full test suite for the model? :-)

Soonhwan-Kwon commented:
> @Soonhwan-Kwon, could you maybe also add a configuration file (just copy the xlm-roberta one) and also add a full test suite for the model? :-)

sorry for the late response. I already added config.json, so is it the tokenizer.json you're talking about? And I added simple test for models(tests/test_modeling_xlm_roberta_xl.py) but where can I find the full test suite?

patrickvonplaten commented:
Hey @Soonhwan-Kwon, 

I meant more a new `configuration_xlm_roberta_xl.py` python file that is more or less a copy of `configuration_xlm_robert.py`:-) But I see that the configs are exactly similar so maybe we can leave as is. 

@sgugger @LysandreJik - This PR adds the checkpoints of https://ai.facebook.com/blog/-xlm-r-state-of-the-art-cross-lingual-understanding-through-self-supervision/ to `transformers`. The model is essentially a "scaled-up" version of https://huggingface.co/docs/transformers/master/en/model_doc/xlmroberta#overview . Since the "scaled-up" version has a significantly different architecture (layer_norm is used very different amongst others) we decided to make a new `modeling_xlm_roberta_xl.py` model file. Now would it be ok for you to a) **not** have a corresponding `configuration_xlm_roberta_xl.py` and just use the `config_xlm_roberta.py` code or do you prefer b) adding a new `configuration_xlm_roberta_xl.py` file for consistency? I'm a bit indifferent here, but do slighly prefer b). What do you think? 

@Soonhwan-Kwon - there are some failing tests which I think can partly be solved by rebasing to current master. Otherwise, if ok for you I'm also happy to dive in the PR and help you finish the last parts of it. Let me know what you prefer :-)

LysandreJik commented:
Hey @Soonhwan-Kwon, thanks a lot for your PR!!

@patrickvonplaten, I prefer `b)`: a lot of the library is built on the assumption that you have one configuration file/object per modeling file/model objects. Since we've authorized auto models to map one configuration to multiple models this isn't as much of an issue as it could have been in the past, but I'm positive we'll find edge cases where it doesn't work as well as we expect it to simply because of the wrong assumption.

sgugger commented:
Also, since it falls into our "new architecture test", there should be a new folder regrouping this modeling file and configuration file instead of putting everything in the xlm-roberta folder.

Soonhwan-Kwon commented:
> Hey @Soonhwan-Kwon,
> 
> I meant more a new `configuration_xlm_roberta_xl.py` python file that is more or less a copy of `configuration_xlm_robert.py`:-) But I see that the configs are exactly similar so maybe we can leave as is.
> 
> @sgugger @LysandreJik - This PR adds the checkpoints of https://ai.facebook.com/blog/-xlm-r-state-of-the-art-cross-lingual-understanding-through-self-supervision/ to `transformers`. The model is essentially a "scaled-up" version of https://huggingface.co/docs/transformers/master/en/model_doc/xlmroberta#overview . Since the "scaled-up" version has a significantly different architecture (layer_norm is used very different amongst others) we decided to make a new `modeling_xlm_roberta_xl.py` model file. Now would it be ok for you to a) **not** have a corresponding `configuration_xlm_roberta_xl.py` and just use the `config_xlm_roberta.py` code or do you prefer b) adding a new `configuration_xlm_roberta_xl.py` file for consistency? I'm a bit indifferent here, but do slighly prefer b). What do you think?
> 
> @Soonhwan-Kwon - there are some failing tests which I think can partly be solved by rebasing to current master. Otherwise, if ok for you I'm also happy to dive in the PR and help you finish the last parts of it. Let me know what you prefer :-)

@patrickvonplaten Sure, I will be glad if you help the last parts and feel free to dive in this PR.

Soonhwan-Kwon commented:
@patrickvonplaten I added you as collaborator in my repo, perhaps you might need access.

patrickvonplaten commented:
Thanks @Soonhwan-Kwon - I'll try to tackle this tomorrow :-)

patrickvonplaten committed:
fix some stuff

patrickvonplaten committed:
Merge branch 'master' of https://github.com/huggingface/transformers into xlm_xl

patrickvonplaten committed:
up

patrickvonplaten committed:
correct init

patrickvonplaten committed:
fix more

patrickvonplaten commented on src/transformers/models/xlm_roberta_xl/configuration_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,154 @@
+# coding=utf-8
+# Copyright 2018 The Google AI Language Team Authors and The HuggingFace Inc. team.
```
change docstring to Hugging Face and correct year

Soonhwan-Kwon replied:
changed as your suggestions

patrickvonplaten commented on src/transformers/models/xlm_roberta_xl/configuration_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,154 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" XLM-RoXLM_ROBERTA_XLa configuration"""
```
typo

Soonhwan-Kwon replied:
fixed typo

patrickvonplaten commented on src/transformers/models/xlm_roberta_xl/configuration_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,154 @@
+}
+
+
+class XLMRobertaXLConfig(PretrainedConfig):
```
The default values need to be changed to https://huggingface.co/Soonhwan-Kwon/xlm-roberta-xlarge/blob/main/config.json

Soonhwan-Kwon replied:
changed default values as config, thank you for the heads up!

patrickvonplaten commented on tests/test_modeling_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,61 @@
+if is_torch_available():
+    import torch
+
+    from transformers import XLMRobertaXLModel
```
We need more tests here. We can more or less copy all the tests from: https://github.com/huggingface/transformers/blob/e68c3756fea7c811d02b8470539ae17ec3ec0e71/tests/test_modeling_roberta.py#L49

Soonhwan-Kwon replied:
Thank you for the review, but the original xlm-roberta also doesn't have full tests like roberta. And I have followed the xlm-roberta test cases.

patrickvonplaten commented on tests/test_modeling_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,61 @@
+
+
+@require_torch
+class XLMRobertaModelIntegrationTest(unittest.TestCase):
```
The tests are passing -> nice!

patrickvonplaten commented on tests/test_modeling_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,61 @@
+class XLMRobertaModelIntegrationTest(unittest.TestCase):
+    @slow
+    def test_xlm_roberta_xlarge(self):
+        model = XLMRobertaXLModel.from_pretrained("Soonhwan-Kwon/xlm-roberta-xlarge")
```
```suggestion
        model = XLMRobertaXLModel.from_pretrained("Soonhwan-Kwon/xlm-roberta-xlarge").to(torch_device)
```

patrickvonplaten commented on tests/test_modeling_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,61 @@
+    @slow
+    def test_xlm_roberta_xlarge(self):
+        model = XLMRobertaXLModel.from_pretrained("Soonhwan-Kwon/xlm-roberta-xlarge")
+        input_ids = torch.tensor([[0, 581, 10269, 83, 99942, 136, 60742, 23, 70, 80583, 18276, 2]])
```
```suggestion
        input_ids = torch.tensor([[0, 581, 10269, 83, 99942, 136, 60742, 23, 70, 80583, 18276, 2]], device=torch_device)
```

patrickvonplaten commented on src/transformers/models/xlm_roberta_xl/modeling_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,1544 @@
+# coding=utf-8
+# Copyright 2022 The HuggingFace Inc. team.
+# Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
```
```suggestion
```

Soonhwan-Kwon replied:
fixed as suggestion

patrickvonplaten commented:
@Soonhwan-Kwon - I've done some changes and left some final To-Dos in case you would like to tackle them :-) 

We should also add this model to the README.md with this paper: https://arxiv.org/pdf/2105.00572.pdf and give it a doc

Feel free to finish those last tasks if you want. Otherwise, I think I'm available to do this by mid- or end of the next week :-)

Soonhwan-Kwon committed:
fix as suggestions

Soonhwan-Kwon committed:
add torch_device

stefan-it commented on src/transformers/models/xlm_roberta_xl/configuration_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,153 @@
+
+
+    Args:
+        vocab_size (`int`, *optional*, defaults to 30522):
```
```suggestion
        vocab_size (`int`, *optional*, defaults to 250880):
```

Soonhwan-Kwon replied:
thank you for the suggestion, I missed docstring, thank you!

stefan-it commented on src/transformers/models/xlm_roberta_xl/configuration_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,153 @@
+        vocab_size (`int`, *optional*, defaults to 30522):
+            Vocabulary size of the XLM_ROBERTA_XL model. Defines the number of different tokens that can be represented
+            by the `inputs_ids` passed when calling [`XLMRobertaXLModel`] or [`TFXLMRobertaXLModel`].
+        hidden_size (`int`, *optional*, defaults to 768):
```
```suggestion
        hidden_size (`int`, *optional*, defaults to 2560):
```

stefan-it commented on src/transformers/models/xlm_roberta_xl/configuration_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,153 @@
+            The dropout probability for all fully connected layers in the embeddings, encoder, and pooler.
+        attention_probs_dropout_prob (`float`, *optional*, defaults to 0.1):
+            The dropout ratio for the attention probabilities.
+        max_position_embeddings (`int`, *optional*, defaults to 512):
```
```suggestion
        max_position_embeddings (`int`, *optional*, defaults to 514):
```

stefan-it commented on src/transformers/models/xlm_roberta_xl/configuration_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,153 @@
+            by the `inputs_ids` passed when calling [`XLMRobertaXLModel`] or [`TFXLMRobertaXLModel`].
+        hidden_size (`int`, *optional*, defaults to 768):
+            Dimensionality of the encoder layers and the pooler layer.
+        num_hidden_layers (`int`, *optional*, defaults to 12):
```
```suggestion
        num_hidden_layers (`int`, *optional*, defaults to 36):
```

stefan-it commented on src/transformers/models/xlm_roberta_xl/configuration_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,153 @@
+            Dimensionality of the encoder layers and the pooler layer.
+        num_hidden_layers (`int`, *optional*, defaults to 12):
+            Number of hidden layers in the Transformer encoder.
+        num_attention_heads (`int`, *optional*, defaults to 12):
```
```suggestion
        num_attention_heads (`int`, *optional*, defaults to 32):
```

stefan-it commented on src/transformers/models/xlm_roberta_xl/configuration_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,153 @@
+            Number of hidden layers in the Transformer encoder.
+        num_attention_heads (`int`, *optional*, defaults to 12):
+            Number of attention heads for each attention layer in the Transformer encoder.
+        intermediate_size (`int`, *optional*, defaults to 3072):
```
```suggestion
        intermediate_size (`int`, *optional*, defaults to 10240):
```

stefan-it commented on src/transformers/models/xlm_roberta_xl/configuration_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,153 @@
+            [`TFXLMRobertaXLModel`].
+        initializer_range (`float`, *optional*, defaults to 0.02):
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        layer_norm_eps (`float`, *optional*, defaults to 1e-12):
```
```suggestion
        layer_norm_eps (`float`, *optional*, defaults to 1e-05):
```

stefan-it commented on src/transformers/models/xlm_roberta_xl/configuration_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,153 @@
+        max_position_embeddings (`int`, *optional*, defaults to 512):
+            The maximum sequence length that this model might ever be used with. Typically set this to something large
+            just in case (e.g., 512 or 1024 or 2048).
+        type_vocab_size (`int`, *optional*, defaults to 2):
```
```suggestion
        type_vocab_size (`int`, *optional*, defaults to 1):
```

stefan-it commented:
Thanks for working on that @Soonhwan-Kwon . I made some minor suggestions and will look at the tokenization part now (to check if there are any differences between XLM-R and XLM-R-XL/XXL :)

Soonhwan-Kwon committed:
fix default values of doc strings

stefan-it commented:
The tokenization part is working as expected. Here are some details:

* Underlying sentence piece models (XLM-R and XLM-R-XL) are identical (checked that via `torch.hub.load`, that downloads the model and stores them under `~/.cache/torch/pytorch_fairseq`). Checksums are the same.

* Tokenizer mapping (fairseq to spm model) is thankfully the same as for XLM-R, here I documented that mapping:

https://github.com/huggingface/transformers/blob/2e9af294940083915ccb2740a7c8d5b154194f15/src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py#L163-L167

* But where does this vocab size "mismatch" come from? XLM-R has a vocab size of 250,002, whereas XLM-R-XL has 250,880.

Fairseq comes with an own dictionary file (`dict.txt`), for XLM-R it has 249,997 entries, and 250,875 entries for XLM-R-XL.

The dictionary file for XLM-R-XL is the same as XLM-R, but it contains `madeupword` tokens ranging from `madeupword0` to `madeupword877` at the end.

patrickvonplaten committed:
fix leftovers

patrickvonplaten committed:
Merge branch 'master' of https://github.com/huggingface/transformers into xlm_xl

patrickvonplaten committed:
merge to master

sgugger commented on src/transformers/models/xlm_roberta_xl/modeling_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,1543 @@
+        return position_ids.unsqueeze(0).expand(input_shape)
+
+
+# Copied from transformers.models.bert.modeling_bert.BertSelfAttention with Bert->Roberta
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertSelfAttention with Bert->XLMRoberta
```

sgugger commented on src/transformers/models/xlm_roberta_xl/modeling_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,1543 @@
+        return hidden_states
+
+
+# Copied from transformers.models.bert.modeling_bert.BertAttention with Bert->Roberta
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertAttention with Bert->XLMRoberta
```

sgugger commented on src/transformers/models/xlm_roberta_xl/modeling_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,1543 @@
+        return hidden_states
+
+
+# Copied from transformers.models.bert.modeling_bert.BertLayer with Bert->Roberta
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertLayer with Bert->XLMRoberta
```

sgugger commented on src/transformers/models/xlm_roberta_xl/modeling_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,1543 @@
+        return layer_output
+
+
+# Copied from transformers.models.bert.modeling_bert.BertEncoder with Bert->Roberta
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertEncoder with Bert->XLMRoberta
```

sgugger reviewed (approved):
Looking good! Careful with some `Copied from statements that haven't been adapted to the model name.

patrickvonplaten committed:
up

patrickvonplaten committed:
correct hub names

patrickvonplaten committed:
fix docs

patrickvonplaten committed:
fix model

patrickvonplaten committed:
up

patrickvonplaten committed:
finalize

patrickvonplaten commented on docs/source/model_doc/xlm-roberta-xl.mdx:
```diff
@@ -0,0 +1,69 @@
+  not require `lang` tensors to understand which language is used, and should be able to determine the correct
+  language from the input ids.
+
+This model was contributed by [Soonhwan-Kwon](https://github.com/Soonhwan-Kwon) and [stefan-it](https://huggingface.co/stefan-it). The original code can be found [here](https://github.com/pytorch/fairseq/tree/master/examples/xlmr).
```
Mentioned you here @Soonhwan-Kwon and @stefan-it

patrickvonplaten reviewed (approved):
**UPDATE**: The PR should be ready for merge now. The checkpoints have been moved to the Facebook's org: https://huggingface.co/models?other=xlm-roberta-xl and added some model cards. @Soonhwan-Kwon, I've made you the "main" contributor for this model.

patrickvonplaten committed:
last fix

patrickvonplaten commented:
@stefan-it - it would also be great if you could do a final review :-)

sgugger commented on src/transformers/models/xlm_roberta_xl/configuration_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,151 @@
+    Args:
+        vocab_size (`int`, *optional*, defaults to 250880):
+            Vocabulary size of the XLM_ROBERTA_XL model. Defines the number of different tokens that can be represented
+            by the `inputs_ids` passed when calling [`XLMRobertaXLModel`] or [`TFXLMRobertaXLModel`].
```
```suggestion
            by the `inputs_ids` passed when calling [`XLMRobertaXLModel`].
```
I don't think there is a TF implementation.

sgugger commented on src/transformers/models/xlm_roberta_xl/modeling_xlm_roberta_xl.py:
```diff
@@ -0,0 +1,1542 @@
+        )
+
+
+def create_position_ids_from_input_ids(input_ids, padding_idx, past_key_values_length=0):
```
Missing a copied from roberta I think.

sgugger reviewed (approved):
Looking good!

patrickvonplaten commented on README.md:
```diff
@@ -322,6 +322,7 @@ AWARE PRE-TRAINING](https://arxiv.org/abs/2110.05752) by Sanyuan Chen, Yu Wu, Ch
 1. **[XLM](https://huggingface.co/docs/transformers/model_doc/xlm)** (from Facebook) released together with the paper [Cross-lingual Language Model Pretraining](https://arxiv.org/abs/1901.07291) by Guillaume Lample and Alexis Conneau.
 1. **[XLM-ProphetNet](https://huggingface.co/docs/transformers/model_doc/xlm-prophetnet)** (from Microsoft Research) released with the paper [ProphetNet: Predicting Future N-gram for Sequence-to-Sequence Pre-training](https://arxiv.org/abs/2001.04063) by Yu Yan, Weizhen Qi, Yeyun Gong, Dayiheng Liu, Nan Duan, Jiusheng Chen, Ruofei Zhang and Ming Zhou.
 1. **[XLM-RoBERTa](https://huggingface.co/docs/transformers/model_doc/xlm-roberta)** (from Facebook AI), released together with the paper [Unsupervised Cross-lingual Representation Learning at Scale](https://arxiv.org/abs/1911.02116) by Alexis Conneau*, Kartikay Khandelwal*, Naman Goyal, Vishrav Chaudhary, Guillaume Wenzek, Francisco Guzmán, Edouard Grave, Myle Ott, Luke Zettlemoyer and Veselin S ...
+1. **[XLM-RoBERTa-XL](https://huggingface.co/docs/transformers/model_doc/xlm-roberta-xl)** (from Facebook AI), released together with the paper [Larger-Scale Transformers for Multilingual Masked Language Modeling](https://arxiv.org/abs/2105.00572) by Naman Goyal, Jingfei Du, Myle Ott, Giri Anantharaman, Alexis Conneau.
```
Gosh, I'm always forgetting this - sorry!

patrickvonplaten committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

patrickvonplaten committed:
add copied from

patrickvonplaten committed:
make style

stefan-it commented:
Really cool! I'm currently running experiments on token classification with that new model :hugs:

Soonhwan-Kwon commented:
@patrickvonplaten @sgugger @stefan-it  Thank you for the merge, it was a great experience, and I came to respect committers of transformers. And below revert is just miss click, sorry.
